File: python/algorithms/LCS.py
""" longest common subsequence """
str1 = input("Enter the string : ")
str2 = input("Enter the string to compare : ")
n = len(str1)+1
m = len(str2)+1

arr = [[0 for j in range(m)] for i in range(n)]

def check(i, j):
    if str1[i] == str2[j] :
        return 1
    else:
        return 0

for i in range(1,n):
    for j in range(1,m):
        if check(i-1,j-1):
            arr[i][j] = arr[i-1][j-1]+1
            lcs += str1[i-1]
        else:
            arr[i][j] = max([arr[i-1][j], arr[i][j-1]])

lcs = ""
i,j = n-1,m-1
while (i!=0 or j!=0):
    if arr[i][j]==arr[i-1][j-1]+1:
        lcs = str1[i-1] + lcs
        i -= 1
        j -= 1
    elif arr[i][j] == arr[i-1][j]:
        i -= 1
    else:
        j -= 1

print("Longest Common Subsequence : "+lcs)
print("Length of Longest Increasing Subsequence : "+str(arr[n-1][m-1]))

File: java/algorithms/LCS.java
import java.util.*;
public class LCS{ 

int lcs( char[] X, char[] Y, int m, int n ) 
{ 
	if (m == 0 || n == 0) 
		return 0; 
	if (X[m-1] == Y[n-1]) 
		return 1 + lcs(X, Y, m-1, n-1); 
	else
		return max(lcs(X, Y, m, n-1), lcs(X, Y, m-1, n)); 
} 

int max(int a, int b) { 
	return (a > b)? a : b; 
} 

public static void main(String[] args) { 
	Scanner sc=new Scanner(System.in);
	LCS lcs = new LCS(); 
	System.out.print("Enter First String :");
	String s1 = sc.next(); 
	System.out.print("Enter Second String :");
	String s2 = sc.next(); 

	char[] X=s1.toCharArray(); 
	char[] Y=s2.toCharArray(); 
	int m = X.length; 
	int n = Y.length; 

	System.out.println("Length of LCS is" + " " + lcs.lcs( X, Y, m, n ) ); 
} 

} 

